File: tools/find_action.py
#!/usr/bin/env python3
"""Search the dumped REAPER action list.

    tools/find_action.py "zoom in horizontal"
    tools/find_action.py --section midi_editor "note: "
    tools/find_action.py --id 40222

All terms must appear (AND, case-insensitive, order-independent).
"""
import argparse
import pathlib
import sys

DUMP = pathlib.Path(__file__).resolve().parent.parent / "data" / "reaper-actions-7.78-linux.tsv"


def load(path=DUMP):
    rows = []
    with open(path, encoding="utf-8", errors="replace") as fh:
        next(fh)  # header
        for line in fh:
            parts = line.rstrip("\n").split("\t")
            if len(parts) >= 5:
                rows.append(
                    {
                        "section": parts[0],
                        "section_id": parts[1],
                        "command_id": parts[2],
                        "named_id": parts[3],
                        "name": parts[4],
                    }
                )
    return rows


def search(rows, terms, section="main"):
    terms = [t.lower() for t in terms]
    out = []
    for r in rows:
        if section and r["section"] != section:
            continue
        hay = r["name"].lower()
        if all(t in hay for t in terms):
            out.append(r)
    return out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("terms", nargs="*")
    ap.add_argument("--section", default="main")
    ap.add_argument("--id", help="look up one command id exactly")
    ap.add_argument("--limit", type=int, default=25)
    args = ap.parse_args()

    rows = load()

    if args.id:
        hits = [r for r in rows if r["command_id"] == args.id and r["section"] == args.section]
    else:
        if not args.terms:
            ap.error("give search terms or --id")
        hits = search(rows, args.terms, args.section)

    for r in hits[: args.limit]:
        print(f"{r['command_id']:>8}  {r['name']}")
    if len(hits) > args.limit:
        print(f"... {len(hits) - args.limit} more")
    if not hits:
        print("(no match)", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/cli/find-action.ts
import { loadActions, ActionIndex } from '@/adapters/reaper/actions'
import { parseFindAction } from '@/cli/args'

export function cmdFindAction(argv: string[]): number {
  const a = parseFindAction(argv)
  const rows = loadActions()
  const idx = new ActionIndex(rows)
  if (a.id) {
    const hit = rows.find((r) => r.commandId === a.id && r.section === (a.section ?? 'main'))
    if (!hit) { console.error(`no action ${a.id}`); return 1 }
    console.log(`${hit.commandId}\t${hit.actionName}`)
    return 0
  }
  for (const r of idx.find(a.terms, a.section)) console.log(`${r.section}\t${r.commandId}\t${r.actionName}`)
  return 0
}
